openai/tests/test_client.py

```diff
@@ -123,8 +172,320 @@ def mock_embeddings_response(*args, **kwargs):
+
+    # valid override
+    kwargs = {
+        "api_base": "expected",
+        "api_key": "expected",
+        "api_version": "expected",
+        "prompt": "expected",
+    }
+
+    overrides = {
+        "temperature": 0.1
+    }
+
+    client._populate_args(kwargs, **overrides)
+
+    assert kwargs == {
+        "api_base": "expected",
+        "api_key": "expected",
+        "api_type": "open_ai",
+        "api_version": "expected",
+        "prompt": "expected",
+        "organization": None,
+        "temperature": 0.1
+    }
+
+
+    # unexpected override by user
+    kwargs = {
+        "prompt": "expected",
+        "api_base": "expected",
+        "api_key": "expected",
+        "api_type": "expected",
+        "api_version": "expected",
+        "organization": "expected",
+        "stream": True
+    }
+
+    overrides = {
+        "stream": False
+    }
+
+    with pytest.raises(TypeError):
+        client._populate_args(kwargs, **overrides)
+
+    # attempt to change api_base on per-method call
+    kwargs = {
+        "prompt": "expected",
+        "api_base": "expected",
+        "api_key": "expected",
+        "api_type": "expected",
+        "api_version": "expected",
+        "organization": "expected",
+        "stream": True
+    }
+
+    overrides = {
+        "api_base": "update",
+    }
+
+    with pytest.raises(TypeError):
+        client._populate_args(kwargs, **overrides)
+
+
+def test_normalize_model():
+    client = openai.client.OpenAIClient(backend="azure", api_base="azurebase")
+
+    # azure: deployment_id --> deployment_id
+    kwargs = {"deployment_id": "ada"}
+    client._normalize_model(kwargs)
+    assert kwargs == {"deployment_id": "ada"}
+
+    # azure: engine --> engine
+    kwargs = {"engine": "ada"}
+    client._normalize_model(kwargs)
+    assert kwargs == {"engine": "ada"}
+
+    # azure: model --> deployment_id (normalized)
+    kwargs = {"model": "ada"}
+    client._normalize_model(kwargs)
+    assert kwargs == {"deployment_id": "ada"}
+
+    client = openai.client.OpenAIClient(backend="openai")
+    # openai: deployment_id --> exception
+    kwargs = {"deployment_id": "ada"}
+    client._normalize_model(kwargs)
+    # incorrect arg raised by library
+    assert kwargs == {"deployment_id": "ada"}
+
+    # openai: engine --> engine
+    kwargs = {"engine": "ada"}
+    client._normalize_model(kwargs)
+    assert kwargs == {"engine": "ada"}
+
+    # openai: model --> model
+    kwargs = {"model": "ada"}
+    client._normalize_model(kwargs)
+    assert kwargs == {"model": "ada"}
+
+    # too many args
+    kwargs = {"model": "ada", "deployment_id": "ada"}
+    with pytest.raises(TypeError):
+        client._normalize_model(kwargs)
+
+
+# LIVE TESTS ------------------------------------------------
+# COMPLETION TESTS
+@pytest.mark.parametrize("api_type", API_TYPE)
+@configure_client
+def test_client_completion(api_type, **kwargs):
+    client = kwargs.pop("client")
```

Comment:
Why not make `client` a keyword-only argument to the test

Reply:
Was a workaround to not make client a pytest.fixture, but instead I just made a client a fixture now 😆